src/replace_variables.cpp

```diff
@@ -0,0 +1,107 @@
+#include "vector"
+
+#include "ir/visitor.h"
+#include "replace_variables.h"
+
+#include "frontends/common/constantFolding.h"
+#include "frontends/common/resolveReferences/resolveReferences.h"
+#include "frontends/p4/createBuiltins.h"
+#include "frontends/p4/directCalls.h"
+#include "frontends/p4/simplify.h"
+#include "frontends/p4/simplifyDefUse.h"
+#include "frontends/p4/unusedDeclarations.h"
+
+namespace P4PRUNER {
+
+const IR::Node *ReplaceVariables::postorder(IR::MethodCallExpression *s) {
+    return s;
+}
+const IR::Node *ReplaceVariables::postorder(IR::Expression *s) {
+    auto expr = getOriginal<IR::Expression>();
+    auto type = typeMap->getType((const IR::Node *)expr, true);
+
+    if (typeMap->isLeftValue(expr)) {
+        return s;
+    }
+    int bits = type->width_bits();
+
+    auto decision = get_rnd_pct();
+    if (decision < 0.5 && type->is<IR::Type_Bits>()) {
+        // INFO(s);
+        auto new_elt = new IR::Constant(new IR::Type_Bits(bits, false), 10);
+        // INFO("Replacing " << s << " with " << new_elt);
+        return new_elt;
+    }
+    return s;
+}
+
+const IR::P4Program *apply_replace(const IR::P4Program *program,
+                                   P4PRUNER::PrunerConfig pruner_conf,
+                                   bool has_applied) {
+    P4::ReferenceMap refMap;
+    P4::TypeMap typeMap;
+    const IR::P4Program *temp;
+
+    std::initializer_list<Visitor *> passes;
+
+    PassManager pass_manager(passes);
+
+    if (not has_applied) {
+        pass_manager.addPasses(
+            {new P4::CreateBuiltins(), new P4::ResolveReferences(&refMap, true),
+             new P4::ConstantFolding(&refMap, nullptr),
+             new P4::InstantiateDirectCalls(&refMap),
+             new P4::TypeInference(&refMap, &typeMap, false)});
+    } else {
+        pass_manager.addPasses(
+            {new P4::ResolveReferences(&refMap, true),
+             new P4::TypeInference(&refMap, &typeMap, false)});
+        if (&typeMap != nullptr) {
+            pass_manager.addPasses({new P4::ClearTypeMap(&typeMap)});
```

Comment:
You can also move this sequence of passes into the compiler pass itself. If it is necessary every time. Also have we checked if clearing the type map is really necessary?

Reply:
Do you mean moving this inside the class? 
Yes, apparently it wasn't and removing it even caused an increase in the reduction percentage

Reply:
Yeah inside the class like this basically: https://github.com/p4lang/p4c/blob/master/frontends/p4/unusedDeclarations.h#L123

Reply:
Right, but our replace_variables is not really a compiler pass right? Due to its iterative nature, hence i thought we cant "convert" it to work like a normal compiler pass?

Reply:
Well we could have written it like one since apply_replace always requires the same passes. But it is fine. We can always do this later on.

Reply:
Okay, alright